mitodl/micromasters
# Added check for profile completeness before visiting dashboard

#### What are the relevant tickets?

closes #286 

#### What's this PR do?

This makes it so that users are directed to fill out unfilled portions of their profile before they can interact with the rest of the dashboard.

There's a utility function (`profileComplete`) defined in `util.js`. This returns `true` if a profile is complete, and else it returns an object with messages about what it is incomplete (in general).

We also have a function to check completion (in `containers/App.js`) which runs on `componentDidMount` and `componentDidUpdate`, and it follows this logic:

```
if (
  good profile &&
  already did TOS &&
  not already in /profile && 
  profile isn't complete
) {
  // do some stuff
}
```

so it plays nicely with the TOS page, if a user has both an incomplete profile and has not yet agreed to the terms of service they will be redirected first to the TOS page, and then after agreeing to that they will be redirected to the profile page.

The `profileComplete` utility function returns something that looks like this:

```
{
  url: "/profile/personal",
  title: "Personal Info",
  text: "Please complete personal info"
}
```

where the `url` is where we want to redirect, and the `title` and `text` fields are used to show the user a dialog box. I added a few new actions and a new reducer to handle this (which should be useful in a more general sense for other dialog boxes we might want to show).

Returning the url here means that we can first check the `personal` tab, and if that isn't valid redirect there, then go on to check the professional tab, etc. 

#### Where should the reviewer start?

I guess with the new reducers / redux stuff changes - this is what I feel least confident about (is `ui` the right name for a top-level key? what would be better?).

#### How should this be manually tested?

Try interacting with it in a couple of different ways. If a field is incomplete in the profile it should be redirecting there in all circumstances (i.e. you shouldn't be able to click away from the profile). It shouldn't redirect to the profile if the user hasn't yet accepted the ToS, however.

#### Any background context you want to provide?

#### Screenshots (if appropriate)

#### What GIF best describes this PR or how it makes you feel?
